### DISCUSSION TITLE
Idea to reduce the limitations of 2D firefights

### DISCUSSION THREAD

**Discussion by Kulik242** [FakeFishGames/Barotrauma#4824] [Ideas and suggestions]
2D imposes some serious limitations on firefights. It may be worth considering that the aiming considers the position of the aiming reticle for what gets hit. If an actor is standing close to the shooter while the shooter's aiming reticle is at the opposing side of the screen the actor in front of the shooter may be considered standing aside from the projectile trajectory.
![](https://i.ibb.co/ZMqQBd7/Sn-mka.png)

...here, the mechanic is safe, the engineer gets hit.

This may be an absolute rule or the firearm skill may play a role and there may be a probability to hit the actor standing close wit a low firearm skill.

**Comment by Sviat** [comment]
Agree, must be some way to ignore people closer to character than some threshold

**Comment by mLuby** [comment]


**Reply by Kulik242** [reply]
Thank, you, I didn't know that.

**Comment by NilanthAnimosus** [comment]
Hmm, due to the very nature of how projectiles are handled, and how sometimes some may want to hit something they either don't fully know the distance of (IE. its off screen) or is close and a snap-judgement, this could be very awkward to implement. plus a submarine is usually (From any I've ever seen interior images of) a very cramped space with minimal room to move.

Assuming this is wanted though and barotrauma submarines have good sizable rooms, it makes sense then to be able to fire without directly hitting who is in front as if leaning to avoid shooting an allied character - there would however need to be indications your having a target though perhaps.

Problem is the game uses ray traced projectiles for hit scan, where the first valid target will usually be the one hit from the weapons barrel position, as well as basing it off item physics for other weapon types such as harpoons and railguns. It would be almost awkward to know if the player means to hit close to the crosshair or in the direction of. and it may seem like adding another layer of artificial accuracy to those (Like myself) who don't aim at a target directly but in the direction of.

A none hitscan projectile (IE. harpoon) would have to know the future of its arc. while a hitscan projectile (IE. revolver) would have to check for every single possible target as far as it can go, and choose the closest one to you mouse cursor to achieve this. It could be doable for hitscan, but none hitscan this sounds very hit and miss on how to do.

And this is not even diving into how it will effect the already existing mechanic of choosing when to shoot as an allied character flees or is stupid near an enemy monster trying to wrench it or if you should try to shoot it anyways and risk taking both out.

**Reply by Sviat** [reply]
Sure, but you mixing many different things in one. Lets simplify.

> how sometimes some may want to hit something they either don't fully know the distance of (IE. its off screen) or is close and a snap-judgement,

Lets assume only **one single use-case**: shooting around target directly in-front of gun

> Problem is the game uses ray traced projectiles for hit scan

If you point me to exact code location would be nice, but i guess it is easy to put proximity check.
if (target is closer than Threshold) AND (crosshair is further than Threshold), roll a dice on Weapon skill if failed -> hit.

> It would be almost awkward to know if the player means to hit close to the crosshair

Icon of crosshair should vary on distance to target, having some kind of uncertainty symbol when below that Threshold and being subject for skill check.

> A none hitscan projectile (IE. harpoon) would have to know the future of its arc. 

No needs, just give it 1-2-X frames of "phase shift" and store frame when they were shot.

> how it will effect the already existing mechanic of choosing when to shoot as an allied character flees

Only if he runs towards shooter

> is stupid near an enemy monster trying to wrench it

Only if they both in front of you

> if you should try to shoot it anyways and risk taking both out

Taking risk by rolling dice for Weapon skill

---
Lore-wise: it is much harder to hit target directly in front of you if it is enemy and moves due to higher angular speed; if target is not moving and friendly - it is harder cause you doing your best to not hit him

**Reply by Kulik242** [reply]
As for the problem of shooting into the dark to hit monsters, this system could only work with human actors.
